modules/tinymce/src/themes/silver/main/ts/ui/core/ChoiceControls.ts

```diff
@@ -0,0 +1,150 @@
+  readonly display: (item: T) => string;
+
+  readonly getCurrent: (editor: Editor) => Optional<T>;
+  readonly setCurrent: (editor: Editor, value: T) => void;
+}
+
+const registerController = <T>(editor: Editor, spec: ControlSpec<T>) => {
+  const getMenuItems = (): Menu.ToggleMenuItemSpec[] => {
+    const options = spec.getOptions(editor);
+
+    // All of the API objects (one for each option)
+    const apis: Record<string, Menu.ToggleMenuItemInstanceApi> = {};
+    // The currently active API object (in a destroyable so that it automatically cleans itself up)
+    const lastApi = Singleton.destroyable();
+
+    const callback = () => {
+      const current = spec.getCurrent(editor);
+      const apiOpt = current.map(spec.normalise).bind((key) => Obj.get(apis, key));
+      apiOpt.fold(
+        // If we don't have a menu item for the current state, make sure we're not highlighting anything
+        lastApi.clear,
+        // If we do have a menu item for the current state, highlight it
+        (api) => {
+          lastApi.set({
+            destroy: () => api.setActive(false)
+          });
+          api.setActive(true);
+        }
+      );
+    };
+
+    editor.on('NodeChange', callback);
+
+    return Arr.map(
+      options,
+      (value, i) => ({
+        type: 'togglemenuitem',
+        text: spec.display(value),
+        onSetup: (api) => {
+          apis[spec.normalise(value)] = api;
+
+          if (i + 1 === options.length) {
+            // run the callback once on startup (on the last option so that we know the apis map has been set up)
+            callback();
+          }
+
+          return () => {
+            // only clean up global things once
+            if (i === 0) {
+              editor.off('NodeChange', callback);
+              lastApi.clear();
+            }
+          };
+        },
+        onAction: () => spec.setCurrent(editor, value)
+      })
+    );
+  };
+
+  editor.ui.registry.addMenuButton(spec.name, {
+    tooltip: spec.text,
+    icon: spec.icon,
+    fetch: (callback) => callback(getMenuItems())
+  });
+
+  editor.ui.registry.addNestedMenuItem(spec.name, {
+    type: 'nestedmenuitem',
+    text: spec.text,
+    getSubmenuItems: getMenuItems
+  });
+};
+
+const lineHeightSpec: ControlSpec<string> = {
+  name: 'lineheight',
+  text: 'Line height',
+  icon: 'line-height',
+
+  getOptions: Settings.getLineHeightFormats,
+  normalise: (input) => Dimension.normalise(input, [ 'fixed', 'relative', 'empty' ]).getOr(input),
+  display: Fun.identity,
+
+  getCurrent: (editor) => Optional.some(editor.queryCommandValue('LineHeight')),
+  setCurrent: (editor, value) => editor.execCommand('LineHeight', false, value)
+};
+
+const languageSpec: ControlSpec<ContentLanguage> = {
+  name: 'language',
+  text: 'Language',
+  icon: 'translate',
+
+  getOptions: Settings.getContentLangugaes,
+  normalise: (input) => Type.isUndefined(input.customCode) ? input.code : `${input.code}/${input.customCode}`,
```

Comment:
This really isn't normalising anything anymore, it's generating a unique lookup key so can we please rename this to reflect that?